RIOT-OS/RIOT
# WIP: Porting the Matter standard to RIOT OS

### Contribution description



- Adds a new example application: **rust-matter-light**
- Application code is written in Rust and implements an "OnOff Light" in [Matter](https://csa-iot.org/all-solutions/matter/) protocol
- Potentially supports all boards which provide the `rust_target` feature (currently only tested on `native64`)
- Uses the external `rs-matter` library, where the Matter protocol is (partly) implemented
- Interfaces with RIOT Socket API using the Module [socket_embedded_nal_async_udp](https://rustdoc.etonomy.org/riot_wrappers/socket_embedded_nal_async_udp/index.html)
- Depends on a [patch of rs-matter](https://github.com/maikerlab/rs-matter/tree/feature/RIOT_OS), which enables the Cargo feature `riot-os` to make it work with RIOT

### Testing procedure



- Follow the guides to build the [chip-tool](https://project-chip.github.io/connectedhomeip-doc/guides/chip_tool_guide.html)
- Build and flash the example on `native64` Target: `BOARD=native64 make all flash term`
- Discover device on network: `./chip-tool discover commissionables`
- QR Code and manual pairing code for commissioning must be printed on terminal
- Commission the device using the manual pairing code (without "-") and arbitrary chosen node id (e. g. `12344321`): `./chip-tool pairing code <node-id> <pairing-code>`

Test supported functionality of commissioned device:
- Read product name: `./chip-tool basicinformation read product-name <node-id> 0`
- Toggle On/Off state: `./chip-tool onoff on <node-id> 1`
- Read On/Off state: `./chip-tool onoff read on-off <node-id> 1`

See [CHIP-Tool Documentation](https://project-chip.github.io/connectedhomeip-doc/guides/chip_tool_guide.html#supported-commands-and-options) for more supported commands.

### Issues/PRs references



Resolves #19929.
